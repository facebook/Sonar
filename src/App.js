/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */

import React from 'react';
import {FlexColumn, FlexRow} from 'sonar';
import {connect} from 'react-redux';
import {toggleBugDialogVisible} from './reducers/application.js';
import WelcomeScreen from './chrome/WelcomeScreen.js';
import SonarTitleBar from './chrome/SonarTitleBar.js';
import MainSidebar from './chrome/MainSidebar.js';
import BugReporterDialog from './chrome/BugReporterDialog.js';
import ErrorBar from './chrome/ErrorBar.js';
import PluginContainer from './PluginContainer.js';
import PluginManager from './chrome/PluginManager.js';

import type Logger from './fb-stubs/Logger.js';
import type BugReporter from './fb-stubs/BugReporter.js';

type Props = {
  logger: Logger,
  bugReporter: BugReporter,
  leftSidebarVisible: boolean,
  bugDialogVisible: boolean,
  pluginManagerVisible: boolean,
  selectedDeviceIndex: number,
  error: ?string,
  toggleBugDialogVisible: (visible?: boolean) => void,
};

export class App extends React.Component<Props> {
  constructor(props: Props) {
    performance.mark('init');
    super(props);
  }

  componentDidMount() {
    this.props.logger.trackTimeSince('init');
  }

  render() {
    return (
      <FlexColumn fill={true}>
        <SonarTitleBar />
        {this.props.bugDialogVisible && (
          <BugReporterDialog
            bugReporter={this.props.bugReporter}
            close={() => this.props.toggleBugDialogVisible(false)}
          />
        )}
        {this.props.selectedDeviceIndex > -1 ? (
          <FlexRow fill={true}>
            {this.props.leftSidebarVisible && <MainSidebar />}
            <PluginContainer logger={this.props.logger} />
          </FlexRow>
        ) : this.props.pluginManagerVisible ? (
          <PluginManager />
        ) : (
          <WelcomeScreen />
        )}
        <ErrorBar text={this.props.error} />
      </FlexColumn>
    );
  }
}

export default connect(
  ({
    application: {pluginManagerVisible, bugDialogVisible, leftSidebarVisible},
    connections: {selectedDeviceIndex},
    server: {error},
  }) => ({
    pluginManagerVisible,
    bugDialogVisible,
    leftSidebarVisible,
    selectedDeviceIndex,
    error,
  }),
  {toggleBugDialogVisible},
)(App);
