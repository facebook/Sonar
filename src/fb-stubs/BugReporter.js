/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */

import LogManager from './Logger';

import type {Store} from '../reducers/index.js';

export default class BugReporter {
  constructor(logManager: LogManager, store: Store) {}
  async report(title: string, body: string): Promise<number> {
    return Promise.resolve(-1);
  }
}
