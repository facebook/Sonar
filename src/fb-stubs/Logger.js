/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */

import type {TrackType, Logger} from '../fb-interfaces/Logger';
import type {Store} from '../reducers/index';
import ScribeLogger from './ScribeLogger';

var instance: ?StubLogger = null;

type Args = {
  isHeadless?: boolean,
};

class StubLogger implements Logger {
  constructor(store: Store, args?: Args) {
    this.scribeLogger = new ScribeLogger(this);
  }

  scribeLogger: ScribeLogger;

  track(type: TrackType, event: string, data: ?any, plugin?: string) {}

  trackTimeSince(mark: string, eventName: ?string) {}

  info(data: any, category: string) {}

  warn(data: any, category: string) {}

  error(data: any, category: string) {}

  debug(data: any, category: string) {}
}

export function init(store: Store, args?: Args): Logger {
  if (instance) {
    throw new Error('Attempted to initialize Logger when already initialized');
  }
  instance = new StubLogger(store);
  return instance;
}

export function getInstance(): Logger {
  if (!instance) {
    throw new Error(
      'Requested Logger instance without initializing it. Make sure init() is called at app start',
    );
  }
  return instance;
}
