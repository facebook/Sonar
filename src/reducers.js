/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 * @format
 */

import type {
  State,
  StatePluginInfo,
  StatePlugins,
  StateClientPlugins,
} from './init.js';
import type {SonarBasePlugin} from 'sonar';
import type Application from './init.js';
import {devicePlugins} from './device-plugins/index.js';
import {SonarPlugin, SonarDevicePlugin} from 'sonar';
import {PluginStateContainer} from './plugin.js';
import BaseDevice from './devices/BaseDevice.js';
import plugins from './plugins/index.js';
import {Client} from './server.js';

const invariant = require('invariant');

type ActivatePluginAction = {|
  appKey: string,
  pluginKey: string,
|};

type TeardownClientAction = {|
  appKey: string,
|};

export function ActivatePlugin(
  app: Application,
  state: State,
  {appKey, pluginKey}: ActivatePluginAction,
) {
  const {activePluginKey, activeAppKey} = state;

  // get currently active plugin
  const activeClientPlugins: ?StateClientPlugins =
    activeAppKey != null ? state.plugins[activeAppKey] : null;
  const activePluginInfo: ?StatePluginInfo =
    activePluginKey != null && activeClientPlugins
      ? activeClientPlugins[activePluginKey]
      : null;

  // check if this plugin is already active
  if (
    activePluginKey === pluginKey &&
    activeAppKey === appKey &&
    activePluginInfo &&
    activePluginInfo.plugin
  ) {
    // this is a noop
    return state;
  }

  // produce new plugins object
  const newPluginsState: StatePlugins = {
    ...state.plugins,
  };

  // check if the currently active plugin needs to be torn down after being deactivated
  if (
    activeAppKey != null &&
    activePluginKey != null &&
    activePluginInfo &&
    activeClientPlugins
  ) {
    const activePlugin: ?SonarBasePlugin<> = activePluginInfo.plugin;
    if (activePlugin && !activePlugin.constructor.persist) {
      // teardown the currently active plugin
      activePlugin._teardown();

      // and remove it's plugin instance so next time it's made active it'll be reloaded
      newPluginsState[activeAppKey] = {
        ...activeClientPlugins,
        [activePluginKey]: {
          plugin: null,
          state: activePluginInfo.state,
        },
      };
    }
  }

  // get the plugin state associated with the new client
  const newClientPluginsState: StateClientPlugins = {
    ...(newPluginsState[appKey] || {}),
  };
  newPluginsState[appKey] = newClientPluginsState;

  // find the Plugin constructor with this key
  let Plugin: Class<SonarBasePlugin<>>;
  for (const FindPlugin of plugins) {
    if (FindPlugin.id === pluginKey) {
      Plugin = FindPlugin;
    }
  }
  for (const FindPlugin of devicePlugins) {
    if (FindPlugin.id === pluginKey) {
      Plugin = FindPlugin;
    }
  }
  invariant(Plugin, 'expected plugin');

  // get target, this could be an app connection or a device
  const clientInfo = state.server.connections.get(appKey);
  let target: Client | BaseDevice;
  if (clientInfo) {
    target = clientInfo.client;
    invariant(
      // $FlowFixMe prototype not known
      Plugin.prototype instanceof SonarPlugin,
      'expected plugin to be an app Plugin',
    );
  } else {
    target = state.devices[appKey];
    invariant(
      // $FlowFixMe prototype not known
      Plugin.prototype instanceof SonarDevicePlugin,
      'expected plugin to be DevicePlugin',
    );
  }
  invariant(target, 'expected target');

  // initialise the client if it hasn't alreadu been
  const thisPluginState: ?StatePluginInfo = newClientPluginsState[pluginKey];
  if (!thisPluginState || !thisPluginState.plugin) {
    const plugin = new Plugin();

    // setup plugin, this is to avoid consumers having to pass args to super
    plugin._setup(target, app);

    // if we already have state for this plugin then rehydrate it
    if (thisPluginState && thisPluginState.state) {
      plugin.state = thisPluginState.state;
    }

    // init plugin - setup broadcasts, initial messages etc
    plugin._init();

    newClientPluginsState[pluginKey] = new PluginStateContainer(
      plugin,
      plugin.state,
    );
  }

  return {
    activeAppKey: appKey,
    activePluginKey: pluginKey,
    plugins: newPluginsState,
  };
}

export function TeardownClient(
  app: Application,
  state: State,
  {appKey}: TeardownClientAction,
) {
  const allPlugins: StatePlugins = {...state.plugins};

  // teardown all plugins
  const clientPlugins: StateClientPlugins = allPlugins[appKey];
  for (const pluginKey in clientPlugins) {
    const {plugin} = clientPlugins[pluginKey];
    if (plugin) {
      plugin._teardown();
    }
  }

  // remove this client
  delete allPlugins[appKey];

  return {
    activeAppKey: state.activeAppKey === appKey ? null : state.activeAppKey,
    activePluginKey:
      state.activeAppKey === appKey ? null : state.activePluginKey,
    plugins: allPlugins,
  };
}
