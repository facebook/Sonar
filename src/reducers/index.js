/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */

import {combineReducers} from 'redux';
import application from './application.js';
import connections from './connections.js';
import server from './server.js';
import pluginStates from './pluginStates.js';
import type {
  State as ApplicationState,
  Action as ApplicationAction,
} from './application.js';
import type {
  State as DevicesState,
  Action as DevicesAction,
} from './connections.js';
import type {
  State as PluginsState,
  Action as PluginsAction,
} from './pluginStates.js';
import type {State as ServerState, Action as ServerAction} from './server.js';
import type {Store as ReduxStore} from 'redux';

export type Store = ReduxStore<
  {
    application: ApplicationState,
    connections: DevicesState,
    pluginStates: PluginsState,
    server: ServerState,
  },
  ApplicationAction | DevicesAction | PluginsAction | ServerAction,
>;

export default combineReducers({
  application,
  connections,
  pluginStates,
  server,
});
