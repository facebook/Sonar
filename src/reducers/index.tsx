/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {combineReducers, Dispatch} from 'redux';
import application, {
  State as ApplicationState,
  Action as ApplicationAction,
} from './application';
import connections, {
  State as DevicesState,
  Action as DevicesAction,
} from './connections';
import pluginStates, {
  State as PluginStatesState,
  Action as PluginStatesAction,
} from './pluginStates';
import notifications, {
  State as NotificationsState,
  Action as NotificationsAction,
} from './notifications';
import plugins, {
  State as PluginsState,
  Action as PluginsAction,
} from './plugins';
import settings, {
  Settings as SettingsState,
  Action as SettingsAction,
} from './settings';
import user, {State as UserState, Action as UserAction} from './user';
import JsonFileStorage from '../utils/jsonFileReduxPersistStorage';
import os from 'os';
import {resolve} from 'path';
import xdg from 'xdg-basedir';
import {persistReducer} from 'redux-persist';
import {PersistPartial} from 'redux-persist/es/persistReducer';

import {Store as ReduxStore, MiddlewareAPI as ReduxMiddlewareAPI} from 'redux';
// @ts-ignore: explicitly need to import index.js, otherwise index.native.js is imported, because redux-persist assumes we are react-native, because we are using metro-bundler
import storage from 'redux-persist/lib/storage/index.js';

export type Actions =
  | ApplicationAction
  | DevicesAction
  | PluginStatesAction
  | NotificationsAction
  | PluginsAction
  | UserAction
  | SettingsAction
  | {type: 'INIT'};

export type State = {
  application: ApplicationState & PersistPartial;
  connections: DevicesState & PersistPartial;
  pluginStates: PluginStatesState;
  notifications: NotificationsState & PersistPartial;
  plugins: PluginsState;
  user: UserState & PersistPartial;
  settingsState: SettingsState & PersistPartial;
};

export type Store = ReduxStore<State, Actions>;
export type MiddlewareAPI = ReduxMiddlewareAPI<Dispatch<Actions>, State>;

const settingsStorage = new JsonFileStorage(
  resolve(
    ...(xdg.config ? [xdg.config] : [os.homedir(), '.config']),
    'flipper',
    'settings.json',
  ),
);

export default combineReducers<State, Actions>({
  application: persistReducer<ApplicationState, Actions>(
    {
      key: 'application',
      storage,
      whitelist: ['flipperRating'],
    },
    application,
  ),
  connections: persistReducer<DevicesState, Actions>(
    {
      key: 'connections',
      storage,
      whitelist: [
        'userPreferredDevice',
        'userPreferredPlugin',
        'userPreferredApp',
        'userLRUPlugins',
      ],
    },
    connections,
  ),
  pluginStates,
  notifications: persistReducer(
    {
      key: 'notifications',
      storage,
      whitelist: ['blacklistedPlugins', 'blacklistedCategories'],
    },
    notifications,
  ),
  plugins,
  user: persistReducer(
    {
      key: 'user',
      storage,
    },
    user,
  ),
  settingsState: persistReducer(
    {key: 'settings', storage: settingsStorage},
    settings,
  ),
});
