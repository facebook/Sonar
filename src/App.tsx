/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import React from 'react';
import {FlexColumn, FlexRow} from 'flipper';
import {connect} from 'react-redux';
import TitleBar from './chrome/TitleBar';
import MainSidebar from './chrome/MainSidebar';
import BugReporterDialog from './chrome/BugReporterDialog';
import ErrorBar from './chrome/ErrorBar';
import ShareSheetExportUrl from './chrome/ShareSheetExportUrl';
import SignInSheet from './chrome/SignInSheet';
import ExportDataPluginSheet from './chrome/ExportDataPluginSheet';
import ShareSheetExportFile from './chrome/ShareSheetExportFile';
import PluginContainer from './PluginContainer';
import Sheet from './chrome/Sheet';
import {ipcRenderer, remote} from 'electron';
import {
  ActiveSheet,
  ShareType,
  ACTIVE_SHEET_BUG_REPORTER,
  ACTIVE_SHEET_PLUGINS,
  ACTIVE_SHEET_SHARE_DATA,
  ACTIVE_SHEET_SIGN_IN,
  ACTIVE_SHEET_SETTINGS,
  ACTIVE_SHEET_SHARE_DATA_IN_FILE,
  ACTIVE_SHEET_SELECT_PLUGINS_TO_EXPORT,
  ACTIVE_SHEET_PLUGIN_SHEET,
} from './reducers/application';
import {Logger} from './fb-interfaces/Logger';
import BugReporter from './fb-stubs/BugReporter';
import {State as Store} from './reducers/index';
import {StaticView, FlipperError} from './reducers/connections';
import PluginManager from './chrome/PluginManager';
import StatusBar from './chrome/StatusBar';
import SettingsSheet from './chrome/SettingsSheet';
const version = remote.app.getVersion();

type OwnProps = {
  logger: Logger;
  bugReporter: BugReporter;
};

type StateFromProps = {
  leftSidebarVisible: boolean;
  errors: FlipperError[];
  activeSheet: ActiveSheet;
  share: ShareType | null;
  staticView: StaticView;
};

type Props = StateFromProps & OwnProps;

export class App extends React.Component<Props> {
  componentDidMount() {
    // track time since launch
    const [s, ns] = process.hrtime();
    const launchEndTime = s * 1e3 + ns / 1e6;
    ipcRenderer.on('getLaunchTime', (_: any, launchStartTime: number) => {
      this.props.logger.track(
        'performance',
        'launchTime',
        launchEndTime - launchStartTime,
      );
    });
    ipcRenderer.send('getLaunchTime');
    ipcRenderer.send('componentDidMount');
  }

  getSheet = (onHide: () => any) => {
    const {activeSheet} = this.props;
    switch (activeSheet) {
      case ACTIVE_SHEET_BUG_REPORTER:
        return (
          <BugReporterDialog
            bugReporter={this.props.bugReporter}
            onHide={onHide}
          />
        );
      case ACTIVE_SHEET_PLUGINS:
        return <PluginManager onHide={onHide} />;
      case ACTIVE_SHEET_SIGN_IN:
        return <SignInSheet onHide={onHide} />;
      case ACTIVE_SHEET_SETTINGS:
        return <SettingsSheet onHide={onHide} />;
      case ACTIVE_SHEET_SELECT_PLUGINS_TO_EXPORT:
        return <ExportDataPluginSheet onHide={onHide} />;
      case ACTIVE_SHEET_SHARE_DATA:
        return (
          <ShareSheetExportUrl
            onHide={onHide}
            logger={this.props.logger}
            closeOnFinish={
              this.props.share != null && this.props.share.closeOnFinish
            }
          />
        );
      case ACTIVE_SHEET_SHARE_DATA_IN_FILE:
        return this.props.share && this.props.share.type === 'file' ? (
          <ShareSheetExportFile
            onHide={onHide}
            file={this.props.share.file}
            logger={this.props.logger}
          />
        ) : (
          (() => {
            console.error('No file provided when calling share sheet.');
            return null;
          })()
        );
      case ACTIVE_SHEET_PLUGIN_SHEET:
        // Currently unused.
        return null;
      default:
        return null;
    }
  };

  render() {
    return (
      <FlexColumn grow={true}>
        <TitleBar version={version} />
        <ErrorBar />
        <Sheet>{this.getSheet}</Sheet>
        <FlexRow grow={true}>
          {this.props.leftSidebarVisible && <MainSidebar />}
          {this.props.staticView != null ? (
            React.createElement(this.props.staticView)
          ) : (
            <PluginContainer logger={this.props.logger} />
          )}
        </FlexRow>
        <StatusBar />
      </FlexColumn>
    );
  }
}

export default connect<StateFromProps, {}, OwnProps, Store>(
  ({
    application: {leftSidebarVisible, activeSheet, share},
    connections: {errors, staticView},
  }) => ({
    leftSidebarVisible,
    activeSheet,
    share: share,
    errors,
    staticView,
  }),
)(App);
