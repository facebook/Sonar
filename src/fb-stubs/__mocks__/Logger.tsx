/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {Store} from '../../reducers/index';
import {getStringFromErrorLike} from '../../utils/errors';
import {Args, Logger, TrackType} from '../../fb-interfaces/Logger';

export function extractError(
  ...data: Array<any>
): {message: string; error: Error} {
  const message = data.map(getStringFromErrorLike).join(' ');
  const error = data.find(e => e instanceof Error) || new Error(message);
  return {
    message,
    error,
  };
}

export class FBLogger implements Logger {
  constructor(_store?: Store, _args?: Args) {}

  track(_type: TrackType, _event: string, _data?: any, _plugin?: string) {}

  trackTimeSince(_mark: string, _eventName?: string) {}

  info = (..._data: Array<any>) => {};

  warn = (..._data: Array<any>) => {};

  error = (..._data: Array<any>) => {};

  debug = (..._data: Array<any>) => {};

  getLogs(): Array<string> {
    return [];
  }
}

export function init(_store: Store, _args?: Args): Logger {
  return new FBLogger();
}

export function getInstance(): Logger {
  return new FBLogger();
}
