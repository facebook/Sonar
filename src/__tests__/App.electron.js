/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */
import React from 'react';
import {App} from '../App.js';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import reducers from '../reducers/index.tsx';
import configureStore from 'redux-mock-store';
import {init as initLogger} from '../fb-stubs/Logger';
import BugReporter from '../fb-stubs/BugReporter.js';

// create redux store with initial state
const mockStore = configureStore([])(reducers(undefined, {type: 'INIT'}));

test('Empty app state matches snapshot', () => {
  const logger = initLogger(mockStore);
  const bugReporter = new BugReporter(logger, mockStore);

  const component = renderer.create(
    <Provider store={mockStore}>
      <App
        logger={logger}
        bugReporter={bugReporter}
        leftSidebarVisible={false}
        selectedDevice={null}
        error={null}
        activeSheet={null}
        share={null}
      />
    </Provider>,
  );
  expect(component.toJSON()).toMatchSnapshot();
});
