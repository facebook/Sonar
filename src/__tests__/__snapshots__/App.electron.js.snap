// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Empty app state matches snapshot 1`] = `
<div
  className="css-1si6n3e"
>
  <div
    className="toolbar css-78eqwq"
  >
    <div
      className="css-q7gju5"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
    >
      <div
        className="css-1h9aj9a"
        color="#acacac"
        size={12}
        src="https://external.xx.fbcdn.net/assets/?name=minus-circle&variant=filled&size=12&set=facebook_icons&density=1x"
      />
      No device selected
    </div>
    <div
      className="css-1ayt83l"
    >
      <div
        className="css-rb77sv"
        disabled={true}
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        title="Take Screenshot"
      >
        <div
          className="css-1dg3xuk"
          color="#acacac"
          size={12}
          src="https://external.xx.fbcdn.net/assets/?name=camera&variant=filled&size=12&set=facebook_icons&density=1x"
        />
      </div>
      <div
        className="css-rb77sv"
        disabled={true}
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        title="Make Screen Recording"
      >
        <div
          className="css-wvm62d"
          color="#acacac"
          size={12}
          src="https://external.xx.fbcdn.net/assets/?name=camcorder&variant=filled&size=12&set=facebook_icons&density=1x"
        />
      </div>
    </div>
    <div
      className="css-12zzrdt"
    />
    <span
      className="css-lli23c"
    >
      4.0.5
      -dev
    </span>
    <div
      className="css-1cecbfb"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      title="Report Bug"
    >
      <div
        className="css-1idwz0j"
        color="#acacac"
        size={12}
        src="https://external.xx.fbcdn.net/assets/?name=bug&variant=filled&size=12&set=facebook_icons&density=1x"
      />
    </div>
    <div
      className="css-1ayt83l"
    >
      <div
        className="css-u1e6jr"
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        title="Toggle Plugins"
      >
        <div
          className="css-2bfl92"
          color="#80a6f5"
          size={20}
          src="icons/sidebar_left.svg"
        />
      </div>
      <div
        className="css-rb77sv"
        disabled={true}
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        title="Toggle Details"
      >
        <div
          className="css-j5i4pm"
          color="#acacac"
          size={20}
          src="icons/sidebar_right.svg"
        />
      </div>
    </div>
  </div>
  <div
    className="css-6xnt2s"
  >
    <div
      id="pluginSheetContents"
      style={
        Object {
          "display": "none",
        }
      }
    />
  </div>
  <div
    className="css-9qtipk"
  >
    <div
      className="css-1atbcdi"
    >
      <div
        className="css-1hs1k35"
      >
        <img
          className="css-1yl4k4p"
          src="./icon.png"
        />
        <span
          className="css-p2pwny"
        >
          Welcome to Flipper
        </span>
        <span
          className="css-1xkgddx"
        >
          Development Mode
        </span>
        <div
          className="css-ls6p57"
          onClick={[Function]}
        >
          <div
            className="css-1hjpcq0"
            color="#8155cb"
            size={20}
            src="https://external.xx.fbcdn.net/assets/?name=rocket&variant=filled&size=20&set=facebook_icons&density=1x"
          />
          <div
            className="css-1k4677w"
          >
            <span
              className="css-1jw94dh"
            >
              Using Flipper
            </span>
            <span
              className="css-bcc6f5"
            >
              Learn how Flipper can help you debug your App
            </span>
          </div>
        </div>
        <div
          className="css-ls6p57"
          onClick={[Function]}
        >
          <div
            className="css-yfo7mj"
            color="#8155cb"
            size={20}
            src="https://external.xx.fbcdn.net/assets/?name=magic-wand&variant=filled&size=20&set=facebook_icons&density=1x"
          />
          <div
            className="css-1k4677w"
          >
            <span
              className="css-1jw94dh"
            >
              Create your own plugin
            </span>
            <span
              className="css-bcc6f5"
            >
              Get started with these pointers
            </span>
          </div>
        </div>
        <div
          className="css-ls6p57"
          onClick={[Function]}
        >
          <div
            className="css-4qdej8"
            color="#8155cb"
            size={20}
            src="https://external.xx.fbcdn.net/assets/?name=tools&variant=filled&size=20&set=facebook_icons&density=1x"
          />
          <div
            className="css-1k4677w"
          >
            <span
              className="css-1jw94dh"
            >
              Add Flipper support to your app
            </span>
            <span
              className="css-bcc6f5"
            >
              Get started with these pointers
            </span>
          </div>
        </div>
        <div
          className="css-ls6p57"
          onClick={[Function]}
        >
          <div
            className="css-l82yfp"
            color="#8155cb"
            size={20}
            src="https://external.xx.fbcdn.net/assets/?name=posts&variant=filled&size=20&set=facebook_icons&density=1x"
          />
          <div
            className="css-1k4677w"
          >
            <span
              className="css-1jw94dh"
            >
              Contributing and Feedback
            </span>
            <span
              className="css-bcc6f5"
            >
              Report issues and help us improve Flipper
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
